### ISSUE TITLE
Inconsistent types between DefineVariable<T> and InquireVariable<T>

### ISSUE THREAD

**Issue by germasch** [ornladios/ADIOS2#1150]
To reproduce (on latest master):
```cxx

#include <adios2.h>

void test1()
{
  using T = char;

  adios2::ADIOS ad(MPI_COMM_WORLD, adios2::DebugON);

  {
    adios2::IO io_writer = ad.DeclareIO("writer");
    auto var = io_writer.DefineVariable<T>("var");
    auto writer = io_writer.Open("test.bp", adios2::Mode::Write);
    
    T test = 99;
    writer.Put(var, test);
    writer.Close();
  }

  {
    adios2::IO io_reader = ad.DeclareIO("reader");
    auto reader = io_reader.Open("test.bp", adios2::Mode::Read);
    auto var = io_reader.InquireVariable<T>("var");
    assert(static_cast<bool>(var));
    
    T test;
    reader.Get(var, test);
    reader.Close();

    assert(test == 99);
  }
}

int main(int argc, char** argv)
{
  MPI_Init(&argc, &argv);
  test1(); 
  MPI_Finalize()
  return 0;
}
```

This test writes a single variable, and then tries to read it back again. The type of the variable can be changed with the `using T = ...` directive. The test succeeds with `using T = int;`, but `using T = char;` fails.

It also fails with `using T = std::size:t` and `using T = unsigned long`.

**Comment by germasch** [comment]
I've tracked this down to `core/IO.tcc`. In `InquireVariable`, thos if statement triggers because of inconsistent types.
```cxx11
    if (itVariable->second.first != helper::GetType<T>())
    {
        return nullptr;
    }
```
In particular, the variable from the file ends up being `signed char` (as is correct on Linux/MacOS), but while `signed char` and `char` behave exactly the same, they are distinct types, and `signed char` != `char`.

This is not pretty, but can be worked around relatively easily by using `signed char` or `unsigned char` for the variable type, though I don't think it's desirable since just `char` is very commonly used.

The `unsigned long` problem is even worse, though. (And `size_t` is an alias for `unsigned long` on 64-bit systems, so that's essentially the ame issue).

The way bp3 currently works, as far as I understand, it's not really possible to recover which actual type was written, because `unsigned long` and `unsigned long long` both map to the same (basically, uint64_t) type id in the file. So when looking in the file, there's no way to know whether the data was one or the other and generate the original type. I suppose the same would happen on a 32-bit system, where unsigned int and unsigned long are indistinguishable.

I think there are really two issues here:
- I don't think it's good that if one writes an `unsigned long` variable and reads it back, the type has changed to `unsigned long long`. However, that's essentially a limitation of, at least, bp3 and can't be changed without introducing new datatypes in the file format, which quite likely opens another can of worms.
- Given that the prior issue isn't likely fixable, it should at least work to read back a variable that's supposedly `unsigned long long` on disk into a variable of type `unsigned long` (on a 64-bit system), which would fix the issue in the test. And it makes some sense, since what's on disk is really uint64_t, which is the representation that unsigned long and unsigned long long use (again, I'm talking about 64-bit systems), so why not allow to read it variables of either type.

I saw some `IsTypeAlias` in the code, which may have been meant to address issues like this, but it's currently unused. Rather than trying to list all kinds of possible aliases for different variables, my suggestion would be to map the types internally to basically the stdint.h types, ie `int{8,16,32,64}_t` and `uint{8,16,32,64}_t`, and then compare those.

Since ull and ul both map to uint64_t, the comparison would give equality, fixing the `T = unsigned long / size_t` issue. Since char and signed char both map to int8_t (on most systems), the `char` problem should also go away.

**Comment by germasch** [comment]
So as for one more update, since I just tried a proof-of-concept of implementation of mapping both sides of the comparison to the stdint types: That makes the comparison succeed, however the next issue is then that there are still separate map per type. In particular, the variable will be listed in the templated-by "signed char" map, but since the InquireVariable function was called for <char>, it'll look in the <char> map and not find the variable.

And for unsigned long, a similar problem happens: adios2 includes 3 types ("unsigned int, unsigned long, unsigned long long" as distinct types, but those are in reality only two distinct types, 32 and 64 bit unsigned, and so if one is looking for a 64 bit unsigned value, one isn't sure where to look -- and it ends up that it's looking in the unsigned long list, but since the file has the type unsigned long long, the var isn't there. While it looks like bp3 has only actually distinct types:
```cxx
       type_unsigned_byte = 50,    //!< type_unsigned_byte
        type_unsigned_short = 51,   //!< type_unsigned_short
        type_unsigned_integer = 52, //!< type_unsigned_integer
        type_unsigned_long = 54,    //!< type_unsigned_long
```
I guess `type_unsigned_long` ends up being mapped to the <unsigned long long> variable list.

Oh well, no easy solutions...

**Comment by williamfgc** [comment]
@germasch we highly encourage to use fixed width types (int8_t, uint64_t) as they map to the proper natural type https://adios2.readthedocs.io/en/latest/components/components.html#variables-data-types for portability. This is an old problem in which the primitive "natural" types (unsigned long, size_t) sizes are not standardized (ranges are), thus making them not portable once we put it in a transport medium (e.g. file system, tape). `unsigned long (Linux)  != unsigned long (Mac/Windows)` same with size_t. It's very confusing.

**Comment by williamfgc** [comment]
One of the challenges is that we can't bypass natural types when using templates....fixed width types are nothing but the correct alias to that particular platform, OS natural type. e.g. `int8_t -> signed char` on Linux

**Comment by williamfgc** [comment]
Also, this is a problem exclusive to the C++11 bindings (templates compile-time checks are not free when dealing with I/O media fixed types). In Python (numpy based), Fortran and C, types passed to adios must be fixed. In fact, we are thinking about removing non-fixed types in our C types enum.

**Comment by germasch** [comment]
Yeah, that all makes sense to me. adios2 internally uses essentially fixed width types already (like the ones from bp3), though it's a bit opaque to the newcomer (me) since they're not actually named that way. And I think that's clearly the right way, if one wants to read an unsigned 64 bit value from an unsigned 64 bit value in a file, that should work no matter how the types are named, which might even change as one moves between systems (like unsigned long does).

After writing the above last night, I had already realized that the reason only I have this trouble is that most other people probably don't use the cxx11 bindings, and in the other cases, the bindings will do specific and unique mapping of the types for them, avoiding the issue.

I do think the documentation isn't clear on that, in fact it mentions
```
char
signed char
unsigned char
short
unsigned short
int
unsigned int
long int
long long int
unsigned long int
unsigned long long int
float
double
long double
std::complex<float>
std::complex<double>
```
as the supported types and then says that the stdint types map to the above, so should also work. I think the way things currently are, and as you say, the supported types should be the `{u,}int{8,16,32,64}_t`, and really, if one uses the native types one may or may not get lucky. It's kinda clear that when mapping between 11 native types and only 8 actually distinct types, a one-to-one mapping isn't possible.

And there's certainly an argument that one shouldn't write `unsigned long` or `size_t` values in the first place, since they change size on different systems. However, in certain cases I think it does make sense, in particular where I'm hitting the problem is in checkpoint/restart, where I'm fine with an underlying assumption that the restart will happen on the same system, and I just want to restore things exactly as they were.  So I would l like to have it so that I can write a `size_t` (a.k.a. `unsigned long`) variable and later read it back, which I currently can't. I also would like to write a `char` array and read it back, which I'm currently doing as a workaround to not being able to write an array of strings (which I believe is not supported).

So I think a possible solution is at the cxx11 bindings level, doing something similar to what the other bindings to, that is, map types as necessary. In particular, map `char` to either `signed char` or `unsigned char` depending on the system. On a 64-bit system, map both `unsigned long` and `unsigned long long` to the uint64_t core type, which will enable that same value to be read back by an `InquireVariable` with either `<unsigned long>`, `<unsigned long long>` or `<uint64_t>` on a 64-bit system, which makes sense.

**Comment by williamfgc** [comment]
@germasch great points, we had several discussion on the topic and we basically reached the same conclusions.  Let me give you some technical answers as we tried things in the past.

- `I do think the documentation isn't clear on that` point taken, I will change it with your suggested comments. There is a box right after the types that talks about the use of fixed-width types for portability. That being said I'd make it more clear in the documentation.
- `However, in certain cases I think it does make sense, in particular where I'm hitting the problem is in checkpoint/restart, where I'm fine with an underlying assumption that the restart will happen on the same system,`  it's difficult for us as a library to predict if a file will be read on the same system 100% of the time. On the other hand, do we really want to encourage people using `size_t `as a type for fixed type media transports? I'd pick the appropriate fix width type for your platform (uint64_t Linux)
- `In particular, map `char` to either `signed char` or `unsigned char` depending on the system` since the C++ standard is ambiguous on char (it only defines 3 different types) compilers historically pick `signed` or `unsigned` see https://www.olcf.ornl.gov/for-users/system-user-guides/summit/summit-user-guide/#compilers . I'd use uint8_t or int8_t for portability, `char` is one of the least portable types.
- `On a 64-bit system, map both `unsigned long` and `unsigned long long` to the uint64_t core type, which will enable that same value to be read back by an `InquireVariable` with either `<unsigned long>`, `<unsigned long long>` or `<uint64_t>` on a 64-bit system, which makes sense.`   We can't do this..for example the data given at Put for an `unsigned long`  Variable on Mac/Windows must be 32-bit , while on Linux it must be 64-bit....remember uint64_t is an alias to a primitive type, not the other way around.
- Unfortunately, all these non-portable media types (`char`, `long`, `size_t`) is decades old C legacy that C++ users need to live with. More modern languages learned from this and have a better approach (e.g. numpy). In C we can control enums, in C++ primitive types are the fundamental types and ultimately no matter what we do/try template types will converge to a primitive one, not a fixed-width type.

**Comment by germasch** [comment]
To start out with, I emphatically agree that the fundamental C/C++ types are a legacy mess, but obviously we aren't going to change that.

In an ideal world, people would use just the stdint types they actually need, though I guess there is still a real need for size_t, as, e.g, it is the argument to `malloc` for a reason.  (In fact, even adios2 is using size_t a lot, though one could even argue about some of that. Using size_t as a basis of Dims, means that  adios2 on a 32 bit system isn't able to handle arrays with > 4G elements, I would guess.)

But, in real life, people use size_t, and char, and unsigned long, so it would be nice to support them, though with the understanding that these are not portable across different systems, so one really shouldn't use them in regular data files, but I think their use in, say, a checkpointing context is appropriate.

Anyway, it turns out that there is already a mapping mechanism between fundamental types and the stdint types by means of `adios2::TypeInfo` -- and it's free to use because it's at compile time.
I used `TypeInfo` in a proof-of-concept mapping between the cxx11 binding and the core adios2. I'll make a PR for it, not because I think it's ready to be merged, but I think it shows that it's feasible to extend the interface such that the same type can be used to read back a variable that was written as a certain type.

**Comment by williamfgc** [comment]
Remember I am speaking from the point of view of an I/O library. When you say "real-life" and cite uses of `size_t`, `char` and `unsigned long` they are all strictly related to operations on heap memory which is volatile (STL sizes, malloc, etc.), not for storage or transport interfaces. Once you move things to any stream (through serialization or buffering), e.g. file, you have to establish a fix size for each record which none of those types do in a portable way (they is even compiler options to select char signness). `size_t` is perfectly reasonable  for `Dims` as they are all memory operations, but in the internal buffer/stream and transport layers we must pick a large enough fixed-width type like `uint64_t`. When talking about data, we must respect the memory alignment, if you write unsigned long in Macs, we can't assume/enforce they are 64-bit.   Now, I am curious about your adios2::TypeInfo PR, we abandoned that idea because backtrace inside the compiler and you will always convergence to a primitive type, not a fixed-width type. What we can do is provide our own set of alias types (adios2::type::int8 like numpy or the C bindings, but std:: is already providing fixed width types), but ultimately any solution to a template type will always be alias of some non-portable primitive. All we can do is promote good practices, we can't restrict primitive template types. Again, any solution will imply trade-offs.

**Comment by williamfgc** [comment]
Not to mention, we are using more an more cloud-based services, so a portable client/server communication is the only way to do. Messing with internal primitives types ask for more trouble than anything when it comes to portability. `size_t` is known for being a non-portable media type for ages.

**Comment by germasch** [comment]
[This comes after my comment that goes with the PR #1154, so I'll try not repeat myself.]

I totally agree that `size_t` has no place in a portable file, and that in general data types in a portable file format need to be unambiguous, that is known width and signedness (which is why I don't like it too much to have bpls show stuff like `int` rather than `int32_t`, but I believe what it shown is for historic reasons, with those strings actually referring to types of well-defined width and signedness), and they probably also should be unique (ie., no two types for the same width / signedness), though I think uniqueness at least has some room for discussion depending on purpose. Anyway, I'm perfectly happy with the internal data model in, e.g., bp3.

I actually find it confusing that the adios2 core does not have uniqueness, ie, `core::Variable<unsigned long>` and `core::Variable<unsigned long long>` both exist and are distinct even though in the end they refer to an undistinguishable representation in a bp3 file. With what I'm proposing in the PR, I believe things in core could be instantiated only for the 8 stdint types, rather than the current 11 fundamental types, and I think that would be preferable.

This is also an aside note, and I'm not proposing to change this, but the choice to base Dims on size_t has drawbacks. In particular, a large data file may not be able to be read on a 32-bit system. I'm not that there is still any relevant use of 32 bit systems left (I'm certainly not aware of any), but it's imaginable that someone would want to run some analysis on a 32-bit Linux system (with large file support) that wants to read a subset of an array from a bp3 file, but would run into trouble because that array has a dimension with > 4G elements.

> When talking about data, we must respect the memory alignment, if you write unsigned long in Macs, we can't assume/enforce they are 64-bit, when it was allocated with 32-bit alignment.

I'm not sure I follow this one, on Mac/Linux-64, unsigned long will be aligned on 64 bits, just like unsigned long long. You won't get misaligned unsigned long's unless you do evil thing with pointers/casts, and that's no different from unsigned long long.

> Now, I am curious about your adios2::TypeInfo PR, we abandoned that idea because backtrace inside the compiler and you will always converge to a primitive type, not a fixed-width type. What we can do is

So hopefully the PR will explain what it does. Yes, things will map to primitive types in the end, no way around that (well, unless you define class-wrapped integer types, which would still behave the same as the internal primitive type, but that would be a lot of hassle with no real benefit other than showing as a distinct type in a backtrace). But what the PR does is to map, e.g., any 64-bit unsigned type to the primitive type underlying `uint64_t`, whatever that is on a given system (could be either `unsigned long` or `unsigned long long` on 64-bit, but whatever it ends up being, the mapped-to type will be the same for any 64 bit unsigned type, primitive or aliases. So at that point, the core doesn't have to provide implementations for both `unsigned long` and `unsigned long long` because we don't know which type might come down, it can just instantiate the template for `uint64_t` because that (or rather the underlying primitive type) is the only that'll be used from the cxx11 interface. (Actually, I haven't looked at the other bindings, and whether they expect to use `core::Variable` on types other than the 8 stdint types. I'd expect the fortran/python interfaces to use the stdint types, but the C binding might well also use the fundamental C types, basically no different from how the cxx11 binding does it right now. So that would prevent to cut down core to just the 8 stdint types at this time, but I don't see why the C bindings couldn't also map thing to the 8 stdint type internally.)

**Comment by williamfgc** [comment]
Good points, your PR is basically on Linux 64-bit:
unsigned long long int, unsigned long int, size_t    reinterpret_cast ---> int64_t  = unsigned long int.

```I actually find it confusing that the adios2 core does not have uniqueness, ie, core::Variable<unsigned long> and core::Variable<unsigned long long> both exist and are distinct even though in the end they refer to an undistinguishable representation in a bp3 file.```
Keep in mind after your PR it's still non-symmetric and non-unique in the sense writing a `<unsigned long int>` can be read with a `<unsigned long long int> or <uint64_t>` not that symmetry with non-portable types is something we want to achieve.   Also,  core::Variable<T> is internal and reused for read and write so it must be able to take any actual type (even if not used at read). We are mixing memory operations for which Variable<T> is for, with buffer representation bp3 types (fixed)

```but the choice to base Dims on size_t has drawbacks. In particular, a large data file may not be able to be read on a 32-bit system. ...``` this goes back to looking at the trade-offs of what makes sense to support and what not in a project.....size_t is fine for STL memory operations, (e.g. size() of a vector in 1D ). At some point we used uint64_t as a prototype and the number of forced static_cast operations was too large and confusing.

```I'd expect the fortran/python interfaces to use the stdint types, but the C binding might well also use the fundamental C types, basically no different from how the cxx11 binding does it right now.```
Fortran, C, Python is all runtime, only C++ allows compile-time type checks. Hence we can limit the types easily in the former languages.

**Comment by williamfgc** [comment]
> When talking about data, we must respect the memory alignment, if you write unsigned long in Macs, we can't assume/enforce they are 64-bit, when it was allocated with 32-bit alignment.

```I'm not sure I follow this one, on Mac/Linux-64, unsigned long will be aligned on 64 bits, just like unsigned long long. You won't get misaligned unsigned long's unless you do evil thing with pointers/casts, and that's no different from unsigned long long.```

My bad, I meant Windows....we support Windows, depite not being POSIX compliant, our CI has Windows. These days with the WSL (Windows subsystem for Linux) Linux and Windows share the same file system on personal computers.

**Comment by germasch** [comment]
Okay, so I'll mostly plead ignorance on Windows. From google, it looks like unsigned long on Windows is 32-bit wide, no matter whether in 32 or 64 bit code. So I suppose 32 bit alignment is fine for it.

Also from google,
> [for malloc] (In Visual C++, the fundamental alignment is the alignment that's required for a double, or 8 bytes. In code that targets 64-bit platforms, it’s 16 bytes.) 

So it looks like no problem on that, either. However, there is certainly a possibility that there are weird systems out there that break things. (Sticking with my Mac/Linux-64 example) one cannot `static_cast` from `unsigned long *` to `unsigned long long *`, one has to use `reinterpret_cast` since they are distinct types. That has a potential of being unsafe. I suppose that there could be systems out there where the types actually encode different behavior, like `unsigned long` only requires 32-bit alignment, but `unsigned long long` 64-bit. I remember there being `movaps` vs `movups` SSE2 instructions, were one requires 16-byte alignment but not the other. ADIOS2 doesn't care about these types, but I guess it goes to show that someone could be crazy and use types with identical representation but differing alignment requirements.

Now, `reinterpret_cast` is already used throughout the code for types of identical width and signedness, so I don't think it has caused any troubles, and it's not likely to, but I can certainly see that it's a potential issue.

*Side note -- which should probably go to the PR, but it fits the topic*. 

Right now, I'm casting, e.g,
`std::vector<unsigned long long>&` to `std::vector<unsigned long>&` in #1154, and that's definitely not guaranteed to be safe. I can't think of any good reason why any STL implementer would specialize `std::vector` for those types differently, considering that they both have the same representation (still talking 64 bit systems here). But maybe someone did... (I know there's usually a special implementation only for std::vector<bool>, which can be hassle enough.) Anyway, that's just something to keep in mind which could need further work in the PR, together with `GetType<>()`.

**Comment by germasch** [comment]
Closed by #1154.
